Fix the openamp_rsc_table sample by specifying the IPM capability for maximum data transfer size

arnopo opened the pull request:
The samples/subsys/ipc/openamp_rsc_table sample requests to transfer data for mailboxes that support data transfer.
Some existing drivers return an error if size is non-null or print a warning message on each transfer.
Update these drivers to be compatible with the subsys/ipc/openamp_rsc_table sample, using LOG_WRN_ONCE to indicate only the first time that the driver does not support data transfers.

This fixes issue introduced in https://github.com/zephyrproject-rtos/zephyr/pull/85617/commits/e50a2ba5217d72a94169c2060bb5d1fb7e19bb1c

erwango commented on drivers/ipm/ipm_stm32_hsem.c:
```diff
@@ -84,10 +84,8 @@ int stm32_hsem_mailbox_ipm_send(const struct device *dev, int wait, uint32_t id,
-		LOG_WRN("stm32 HSEM not support data transfer");
-		return -EMSGSIZE;
-	}
+	if (size)
+		LOG_WRN_ONCE("HSEM driver does not support data transfer over IPM");
```
I'm surprised CI didn't complain but please add `{}`

erwango commented on drivers/ipm/ipm_stm32_ipcc.c:
```diff
@@ -159,9 +159,8 @@ static int stm32_ipcc_mailbox_ipm_send(const struct device *dev, int wait,
-	if (size) {
-		return -EMSGSIZE;
-	}
+	if (size)
+		LOG_WRN_ONCE("IPCC driver does not support data transfer over IPM");
```
Ditto

erwango reviewed (changes requested):
`ipm_send` api specificies:
```
 * @retval -EMSGSIZE If the supplied data size is unsupported by the driver.
```
Based on this, is that ok not to return an error in the driver?

Alternatively, could the sample first check driver support (be resilient to a first error EMSGSIZE received) before actually trying to send data ?

arnopo commented:
> `ipm_send` api specificies:
> 
> ```
>  * @retval -EMSGSIZE If the supplied data size is unsupported by the driver.
> ```
> 
> Based on this, is that ok not to return an error in the driver?

For me yes.  This is the way of working in Linux kernel.

> 
> Alternatively, could the sample first check driver support (be resilient to a first error EMSGSIZE received) before actually trying to send data?

That would mean that the sample takes into account drivers specificity. I would prefer to keep the sample simple.
 
That said if your concern is about HSEM driver, I can remove HSEM update. the driver is not used by the subsys/ipc/openamp_rsc_table sample.

erwango commented:
> For me yes. This is the way of working in Linux kernel.

Ok.

> That would mean that the sample takes into account drivers specificity. I would prefer to keep the sample simple.

My issue is that this change will prevent application to detect driver specificity. So yes user will be informed (if subsystem logs are enabled), but application won't be able to detect issue (underlying hw is not able to transmit data as expected) and report an error to higher levels.
If the issue is that ywe need to keep the sample simple run w/o errors, then it would be better to find ways (sample.yaml filters for instance) to not run it on HW that don't support the feature.

> That said if your concern is about HSEM driver, I can remove HSEM update.

No need, I prefer we find a consistent behavior across drivers.

arnopo commented:
> > For me yes. This is the way of working in Linux kernel.
> 
> Ok.
> 
> > That would mean that the sample takes into account drivers specificity. I would prefer to keep the sample simple.
> 
> My issue is that this change will prevent application to detect driver specificity. So yes user will be informed (if subsystem logs are enabled), but application won't be able to detect issue (underlying hw is not able to transmit data as expected) and report an error to higher levels. If the issue is that ywe need to keep the sample simple run w/o errors, then it would be better to find ways (sample.yaml filters for instance) to not run it on HW that don't support the feature.

There are (at least) two types of mailbox:
- mailbox that works as a dorbell ( non data transmitted only an inter processor notification)
- mailbox that allow to transmit and receive data with a notification

We should support both. 

for that an alternative could be to add an IPM config to indicate the type of  IPM used.

- in `drivers/ipm/Kconfig`

```
config IPM_DATA_TRANSFER
	bool "IPM data transfer support"
	default y
	help
	  Enable the the support of data transmission and reception by the IPM peripheral. 
```

 In `subsys/ipc/openamp_rsc_table sample`
```diff
 int mailbox_notify(void *priv, uint32_t id)
{
	ARG_UNUSED(priv);

	LOG_DBG("%s: msg received", __func__);
-	ipm_send(ipm_handle, 0, id, &id, 4);
+	if (!IS_ENABLED(CONFIG_IPM_DATA_TRANSFER)) {
+		ipm_send(ipm_handle, 0, id, NULL, 0);
+	} else {
+		ipm_send(ipm_handle, 0, id, &id, 4);

	return 0;
}

```
Does this implementation align with your expectations?

erwango commented:
> Does this implementation align with your expectations?

Yes, fine for me. Thanks!

arnopo committed:
driver: ipm: introduce IPM_MAX_DATA_SIZE build configuration

The ipm_send() allows transferring data through the IPM device.
However, depending on the platform, the mailbox peripheral may either
transmit a limited amount of data or not transfer data at all.
Introducing this configuration allows exposing the IPM capability to the
application.
This commit defines a default value of 1024 bytes to avoid impacting
existing IPM drivers and sets the value to 0 for the STM32 IPCC and
STM32 HSEM devices.

This allows, in a next step, other devices to use this configuration to
expose their capabilities instead of using proprietary configuration or
a static definition.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>

arnopo committed:
samples: ipc: openamp_rsc_table: conditional virtio ID sent on IPM device

Some IPM devices return an error if we request to transfer data.
Use IPM_MAX_DATA_SIZE to determine if the virtio ID can be transferred
through the IPM device.
Send data to IPM only if CONFIG_IPM_MAX_DATA_SIZE is not zero.

Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>

arnopo commented:
Finally, I replaced the `IPM_DATA_TRANSFER` configuration with `IPM_MAX_DATA_SIZE`, which can be reused later by other IPM drivers to specify their capabilities in terms of data transfer.

arnopo commented:
@anangl , @masz-nordic:  Could you review please

arnopo commented:
@anangl , @masz-nordic: Gentle reminder

anangl commented on drivers/ipm/Kconfig:
```diff
@@ -8,6 +8,14 @@ menuconfig IPM
 if IPM
 
+config IPM_MAX_DATA_SIZE
+	int "Max data Size supported"
```
```suggestion
	int "Max data size supported"
```

anangl commented on drivers/ipm/Kconfig:
```diff
@@ -8,6 +8,14 @@ menuconfig IPM
+	default 0 if IPM_STM32_IPCC || IPM_STM32_HSEM
+	default 1024
+	help
+	  Define the max size (in byte) of data that can be transmitted or
```
```suggestion
	  Define the max size (in bytes) of data that can be transmitted or
```

arnopo commented:
@anangl : typo fixed + rebase on main branch
